source/shared/cpp/ObjectModel/TextBlock.cpp

```diff
@@ -165,3 +174,191 @@ void TextBlock::SetHorizontalAlignment(const HorizontalAlignment value)
+    std::smatch matches;
+    std::string text = m_text;
+    std::ostringstream parsedostr;
+    time_t offset = 0;
+    bool isDate = false, isValidForm = false; 
+    int  formatStyle = 0;
+    enum MatchIndex
+    {
+        IsDate = 2,
+        Year = 4,
+        Month,
+        Day,
+        Hour,
+        Min,
+        Sec,
+        TimeZone = 12,
+        TZHr,
+        TZMn,
+        Format,
+        Style,
+    };
+    int factor = 0, hours = 0, minutes = 0;
+    struct tm parsedTm = { 0 };
+    int *addrs[] = {&parsedTm.tm_year, &parsedTm.tm_mon,
+        &parsedTm.tm_mday, &parsedTm.tm_hour, &parsedTm.tm_min,
+        &parsedTm.tm_sec, &hours, &minutes};
+    std::vector<int> indexer = {Year, Month, Day, Hour, Min, Sec, TZHr, TZMn};
+
+    while (std::regex_search(text, matches, pattern))
+    {
+        // Date is matched
+        isDate = matches[IsDate].matched;
+
+        if(matches[Style].matched)
+        {
+            // match for long/short/compact
+            formatStyle = matches[Format].str()[2];
+        }
+
+        parsedostr << matches.prefix().str();
+
+        if(!isDate && formatStyle)
+        {
+            parsedostr << matches[0].str();
+            text = matches.suffix().str();
+            continue;
+        }
+
+        for (unsigned int idx = 0; idx < indexer.size(); idx++)
+        {
+            if (matches[indexer[idx]].matched)
+            {
+                // get indexes for time attributes to index into conrresponding matches
+                // and covert it to string
+                *addrs[idx] = stoi(matches[indexer[idx]]);
+            }
+        }
+
+        // check for date and time validation
+        if (parsedTm.tm_mon <= 12 && parsedTm.tm_mday <= 31 && parsedTm.tm_hour <= 24 && 
+            parsedTm.tm_min <= 60 && parsedTm.tm_sec <= 60 && hours <= 24 && minutes <= 60)
+        { 
+            if (parsedTm.tm_mon == 4 || parsedTm.tm_mon == 6 || parsedTm.tm_mon == 9 || parsedTm.tm_mon == 11)
+            { 
+                if (parsedTm.tm_mday <= 30)
+                {
+                    isValidForm = true;
+                }
+            } 
+            else if (parsedTm.tm_mon == 2)
+            {
+                /// check for leap year
+                if ((parsedTm.tm_year % 4 == 0 && parsedTm.tm_year % 100 != 0) || parsedTm.tm_year % 400 == 0)
+                {
+                    if (parsedTm.tm_mday <= 29)
+                    {
+                        isValidForm = true;
+                    }
+                }
+                else if (parsedTm.tm_mday <= 28)
+                {
+                    isValidForm = true;
+                }
+            }
+            else
+            {
+                if (parsedTm.tm_mday <= 31)
+                {
+                    isValidForm = true;
+                }
+            }
+        }
```

Comment:
[nit] consider extracting this section into its own operation. ValidateDateTime or something like it.

Reply:
o.k.


---
In reply to: [150997368](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r150997368) [](ancestors = 150997368)